src/Algebra/Reasoning/Semigroup.agda

```diff
@@ -0,0 +1,157 @@
+  assoc²βε = trans (assoc a (b ∙ c) d) (∙-cong refl (assoc b c d))
+
+  assoc²γβ : (a ∙ b) ∙ (c ∙ d) ≈ (a ∙ (b ∙ c)) ∙ d
+  assoc²γβ = trans (sym (assoc (a ∙ b) c d)) (∙-cong (assoc a b c) refl)
+
+  assoc²γδ : (a ∙ b) ∙ (c ∙ d) ≈ a ∙ ((b ∙ c) ∙ d)
+  assoc²γδ = begin
+    (a ∙ b) ∙ (c ∙ d) ≈⟨ assoc a b (c ∙ d) ⟩
+    a ∙ (b ∙ (c ∙ d)) ≈⟨ ∙-cong refl (sym (assoc b c d)) ⟩
+    a ∙ ((b ∙ c) ∙ d) ∎
+
+  assoc²δα : a ∙ ((b ∙ c) ∙ d) ≈ ((a ∙ b) ∙ c) ∙ d
+  assoc²δα = sym (trans (∙-cong (assoc a b c) refl) (assoc a (b ∙ c) d))
+
+  assoc²δγ : a ∙ ((b ∙ c) ∙ d) ≈ (a ∙ b) ∙ (c ∙ d)
+  assoc²δγ = begin
+    a ∙ ((b ∙ c) ∙ d) ≈⟨ ∙-cong refl (assoc b c d) ⟩
+    a ∙ (b ∙ (c ∙ d)) ≈⟨ sym (assoc a b (c ∙ d)) ⟩
+    (a ∙ b) ∙ (c ∙ d) ∎
+
+  assoc²εα : a ∙ (b ∙ (c ∙ d)) ≈ ((a ∙ b) ∙ c) ∙ d
+  assoc²εα = sym (trans (assoc (a ∙ b) c d) (assoc a b (c ∙ d)))
+
+  assoc²εβ : a ∙ (b ∙ (c ∙ d)) ≈ (a ∙ (b ∙ c)) ∙ d
+  assoc²εβ = sym (trans (assoc a (b ∙ c) d) (∙-cong refl (assoc b c d)))
+
+open Assoc4 public
+
+module Pulls (ab≡c : a ∙ b ≈ c) where
+    pullʳ : ∀ {x} → (x ∙ a) ∙ b ≈ x ∙ c
+    pullʳ {x = x} = begin
+        (x ∙ a) ∙ b ≈⟨ assoc x a b ⟩
+        x ∙ (a ∙ b) ≈⟨ ∙-cong refl ab≡c ⟩
+        x ∙ c       ∎
+
+    pullˡ : ∀ {x} → a ∙ (b ∙ x) ≈ c ∙ x
+    pullˡ {x = x} = begin
+        a ∙ (b ∙ x) ≈⟨ sym (assoc a b x) ⟩
+        (a ∙ b) ∙ x ≈⟨ ∙-cong ab≡c refl ⟩
+        c ∙ x       ∎
+
+    pull-first : ∀ {x y} → a ∙ ((b ∙ x) ∙ y) ≈ c ∙ (x ∙ y)
+    pull-first {x = x} {y = y} = begin
+      a ∙ ((b ∙ x) ∙ y) ≈⟨ ∙-cong refl (assoc b x y) ⟩
+      a ∙ (b ∙ (x ∙ y)) ≈⟨ pullˡ ⟩
+      c ∙ (x ∙ y)       ∎
+
+    pull-center : ∀ {x y} → x ∙ (a ∙ (b ∙ y)) ≈ x ∙ (c ∙ y)
+    pull-center {x = x} {y = y} = ∙-cong refl (pullˡ)
+
+    -- could be called pull₃ʳ
+    pull-last : ∀ {x y} → (x ∙ (y ∙ a)) ∙ b ≈ x ∙ (y ∙ c)
+    pull-last {x = x} {y = y} = begin
+      (x ∙ (y ∙ a)) ∙ b ≈⟨ assoc x (y ∙ a) b ⟩
+      x ∙ ((y ∙ a) ∙ b) ≈⟨ ∙-cong refl (pullʳ {x = y}) ⟩
+      x ∙ (y ∙ c)       ∎
+
+open Pulls public
+
+module Pushes (ab≡c : a ∙ b ≈ c) where
+    pushʳ : x ∙ c ≈ (x ∙ a) ∙ b
+    pushʳ {x = x} = begin
+        x ∙ c       ≈⟨ sym (∙-cong refl ab≡c) ⟩
+        x ∙ (a ∙ b) ≈⟨ sym (assoc x a b) ⟩
+        (x ∙ a) ∙ b ∎
+
+    pushˡ : c ∙ x ≈ a ∙ (b ∙ x)
+    pushˡ {x = x} = begin
+        c ∙ x       ≈⟨ sym (∙-cong ab≡c refl) ⟩
+        (a ∙ b) ∙ x ≈⟨ assoc a b x ⟩
+        a ∙ (b ∙ x) ∎
+open Pushes public
+
+-- operate in the 'center' instead (like pull/push)
+center : a ∙ b ≈ c → (d ∙ a) ∙ (b ∙ e) ≈ d ∙ (c ∙ e)
+center eq = pullʳ (pullˡ eq)
+
+-- operate on the left part, then the right part
+center⁻¹ : a ∙ b ≈ c → x ∙ y ≈ z →  a ∙ ((b ∙ x) ∙ y) ≈ c ∙ z
+center⁻¹ {a = a} {b = b} {c = c} {x = x} {y = y} {z = z} eq eq′ = begin
+  a ∙ ((b ∙ x) ∙ y) ≈⟨ ∙-cong refl (pullʳ eq′) ⟩
+  a ∙ (b ∙ z)       ≈⟨ pullˡ eq ⟩
+  c ∙ z             ∎
+
+push-center : a ∙ b ≈ c → x ∙ (c ∙ y) ≈ x ∙ (a ∙ (b ∙ y))
+push-center eq = sym (pull-center eq)
```

Comment:
So... each of these is parametrised on an assumption of the form `a ∙ b ≈ c`, so why is this not abstracted out as a `module` parameter?

Reply:
```suggestion
-- operate in the 'center' instead (like pull/push)

module _ (a∙b≈c : a ∙ b ≈ c) where

  center : (d ∙ a) ∙ (b ∙ e) ≈ d ∙ (c ∙ e)
  center = pullʳ (pullˡ a∙b≈c)

... etc.
```

Reply:
But see elsewhere about 'better' names!